lib/grist-api.ts

```diff
@@ -0,0 +1,284 @@
+   */
+  public async fetchTable(tableName: string, filters?: IFilterSpec): Promise<IRecord[]> {
+    const query = filters ? `?filter=${encodeURIComponent(JSON.stringify(filters))}` : '';
+    const data: ITableData = await this._call(`tables/${tableName}/data${query}`);
+    // Convert column-oriented data to list of records.
+    debug("fetchTable %s returned %s rows", tableName, data.id.length);
+    return data.id.map((id, index) => mapValues(data, (col) => col[index]));
+  }
+
+  /**
+   * Adds new records to the given table. The data is a list of dictionaries, with keys
+   * corresponding to the columns in the table. Returns a list of added rowIds.
+   */
+  public async addRecords(tableName: string, records: IRecord[]): Promise<number[]> {
+    if (records.length === 0) { return []; }
+
+    const callData: ITableData[] = chunk(records, this._chunkSize).map((recs) => makeTableData(recs));
+
+    const results: number[] = [];
+    for (const data of callData) {
+      debug("addRecords %s %s", tableName, descColValues(data));
+      const resp = await this._call(`tables/${tableName}/data`, data, 'POST');
+      results.push(...(resp || []));
+    }
+    return results;
+  }
+
+  /**
+   * Deletes records from the given table. The data is a list of record IDs.
+   */
+  public async deleteRecords(tableName: string, recordIds: number[]): Promise<void> {
+    // There is an endpoint missing to delete records, but we can use the "apply" endpoint
+    // meanwhile.
+    for (const recIds of chunk(recordIds, this._chunkSize)) {
+      debug("delete_records %s %s records", tableName, recIds.length);
+      const data = [['BulkRemoveRecord', tableName, recIds]];
+      await this._call('apply', data, 'POST');
+    }
+  }
+
+  /**
+   * Update existing records in the given table. The data is a list of objects, with attributes
+   * corresponding to the columns in the table. Each record must contain the key "id" with the
+   * rowId of the row to update.
+   *
+   * If records aren't all for the same set of columns, then a single-call update is impossible,
+   * so we'll make multiple calls.
+   */
+  public async updateRecords(tableName: string, records: IRecord[]): Promise<void> {
+    const groups = new Map<string, IRecord[]>();
+    for (const rec of records) {
+      if (!rec.id || typeof rec.id !== 'number') {
+        throw new Error("updateRecord requires numeric 'id' attribute in each record");
+      }
+      const key = JSON.stringify(Object.keys(rec).sort());
+      const group = groups.get(key) || groups.set(key, []).get(key)!;
+      group.push(rec);
+    }
+
+    const callData: ITableData[] = [];
+    for (const groupRecords of groups.values()) {
+      callData.push(...chunk(groupRecords, this._chunkSize).map((recs) => makeTableData(recs)));
+    }
+
+    for (const data of callData) {
+      debug("updateRecods %s %s", tableName, descColValues(data));
+      await this._call(`tables/${tableName}/data`, data, 'PATCH');
```

Comment:
The "transaction" behavior would certainly be nice, where it's either all or none. But applying a bunch of actions and then rewinding them is not "all or none" -- since there could be other API calls made by other users in the meantime. It's not like a transaction.

Let's keep it simple for now. Transactions are a whole other topic and a big unclear project. I added a comment about it, but I don't think there is an easy TODO for how to change this code.